### PR TITLE
Support Groups

### PR DESCRIPTION
**Pull request by francocorreasosa**
Adds support for Groups (organize applications by groups)
- Home screen:
screen shot 2017-04-11 at 2 04 32 pm

- Group manager:
screen shot 2017-04-11 at 2 05 37 pm

- Group selector in app settings:
screen shot 2017-04-11 at 2 06 05 pm

### CODE REVIEW

File: client/components/Applications/ApplicationsList.jsx [hunk 1]
```diff
@@ -8,39 +8,86 @@ export default class ApplicationsList extends Component {
...
+      
+      return group.apps.length > 0 ? group : null;
+    })
+
+    const allApps = [ ...groupedApps, {
+      name: "Ungrouped apps",
+      apps: ungroupedApps
+    }].filter((group) => group !== null);;
+
+    console.log('allApps', allApps);
```
Comment (sandrinodimattia) [inline client/components/Applications/ApplicationsList.jsx]: Remove this

File: client/components/Applications/ApplicationsList.jsx [hunk 2]
```diff
@@ -8,39 +8,86 @@ export default class ApplicationsList extends Component {
...
+                  const logo = (app.logo) ? app.logo : 'https://cdn.auth0.com/manage/v0.3.1866/img/badge-grey.svg';
+                  const name = app.name || key;
+
+                  const link = app.customURLEnabled ? varstring(app.customURL, {
+                    connection: app.connection,
+                    client_id: app.client,
+                    callback: app.callback
+                  }) : app.loginUrl;
+
+                  console.log('link', link);
```
Comment (sandrinodimattia) [inline client/components/Applications/ApplicationsList.jsx]: Remove this

File: client/components/Applications/ApplicationsList.jsx [hunk 3]
```diff
@@ -8,39 +8,86 @@ export default class ApplicationsList extends Component {
...
-    return nextProps.applications !== this.props.applications;
+    return (nextProps.applications !== this.props.applications) ||
+           (nextProps.groups !== this.props.groups);
   }
 
   render() {
-    const { applications } = this.props;
+    const { applications, groups } = this.props;
+
+    const unpackApp = (appId) => {
```
Comment (sandrinodimattia) [inline client/components/Applications/ApplicationsList.jsx]: Can we not move all of this in the API? Eg, an endpoint returning this:

{
   "Ungrouped Applications": [  xxx ],
   "Group 1": [ ... ]
}

File: server/routes/groups.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import uuid from 'uuid';
+import { Router } from 'express';
+
```
Comment (sandrinodimattia) [inline server/routes/groups.js]: Add unit tests for all these routes so we have 100% coverage

File: server/routes/api.js [hunk 5]
```diff
@@ -37,6 +38,7 @@ export default (storage) => {
     domain: config('AUTH0_DOMAIN')
   });
   api.use('/applications', applications(auth0, storage));
+  api.use('/groups', groups(auth0, storage));
```
Comment (sandrinodimattia) [inline server/routes/api.js]: Let's rename this to "application-groups"

File: server/routes/groups.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import { matchWithApps, saveGroup, deleteGroup } from '../lib/groups';
+
+export default (auth0, storage) => {
+  const api = Router();
+
+  api.get('/', (req, res, next) => {
```
Comment (sandrinodimattia) [inline server/routes/groups.js]: Add the right scopes to all of these routes to restrict access. Look at the applications routes.

File: server/routes/api.js [hunk 7]
```diff
@@ -4,6 +4,7 @@ import { middlewares } from 'auth0-extension-express-tools';
 import config from '../lib/config';
```
Comment (sandrinodimattia) [inline server/routes/api.js]: Add the default scopes for when users and admins login.

File: server/routes/groups.js (hunk 6 above)
Comment (francocorreasosa) [inline server/routes/groups.js]: Should I use the same `read:applications` scope or a different one like `read:groups`?

Comment (francocorreasosa) [inline server/routes/groups.js]: Use `read:application-groups`

**Comment by francocorreasosa** [comment]
The current tests are broken. as I see it they lack of a replacement for the Auth0 Management Middleware.
In the app is instantiated like this:
```js
const auth0 = middlewares.managementApiClient({
  domain: config('AUTH0_DOMAIN')
});
```

But in the tests it is `undefined`.

**Comment by francocorreasosa** [comment]
Reviewing changes. *DO NOT MERGE*

**Pull request closed without merging**